contracts/BillingConnector.sol

```diff
@@ -0,0 +1,155 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.16;
+
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { IBilling } from "./IBilling.sol";
+import { IBillingConnector } from "./IBillingConnector.sol";
+import { Governed } from "./Governed.sol";
+import { ITokenGateway } from "arb-bridge-peripherals/contracts/tokenbridge/libraries/gateway/ITokenGateway.sol";
+
+/**
+ * @title Billing Connector Contract
+ * @dev The billing contract allows for Graph Tokens to be added by a user. The
+ * tokens are immediately sent to the Billing contract on L2 (Arbitrum)
+ * through the GRT token bridge.
+ */
+contract BillingConnector is IBillingConnector, Governed {
+    // -- State --
+    // The contract for interacting with The Graph Token
+    IERC20 private immutable graphToken;
+    // The L1 Token Gateway through which tokens are sent to L2
+    ITokenGateway public l1TokenGateway;
+    // The Billing contract in L2 to which we send tokens
+    address public l2Billing;
+
+    /**
+     * @dev Address of the L1 token gateway was updated
+     */
+    event L1TokenGatewayUpdated(address l1TokenGateway);
+    /**
+     * @dev Address of the L2 Billing contract was updated
+     */
+    event L2BillingUpdated(address l2Billing);
+    /**
+     * @dev Tokens sent to the Billing contract on L2
+     */
+    event TokensSentToL2(address _from, address _to, uint256 _amount);
+    /**
+     * @dev Tokens rescued by the governor
+     */
+    event TokensRescued(address indexed to, address indexed token, uint256 amount);
+
+    /**
+     * @dev Constructor function
+     * @param _l1TokenGateway   L1GraphTokenGateway address
+     * @param _l2Billing Address of the Billing contract on L2
+     * @param _token     Graph Token address
+     * @param _governor  Governor address
+     */
+    constructor(
+        address _l1TokenGateway,
+        address _l2Billing,
+        IERC20 _token,
+        address _governor
+    ) Governed(_governor) {
+        _setL1TokenGateway(_l1TokenGateway);
+        _setL2Billing(_l2Billing);
+        graphToken = _token;
+    }
+
+    /**
+     * @dev Sets the L1 token gateway address
+     * @param _l1TokenGateway New address for the L1 token gateway
+     */
+    function setL1TokenGateway(address _l1TokenGateway) external override onlyGovernor {
+        _setL1TokenGateway(_l1TokenGateway);
+    }
+
+    /**
+     * @dev Sets the L2 Billing address
+     * @param _l2Billing New address for the L2 Billing contract
+     */
+    function setL2Billing(address _l2Billing) external override onlyGovernor {
```

Comment:
This is good as long as `onlyGovernor` is always the Council to avoid anyone to circumvent the gateway allowlist